### DISCUSSION TITLE
Nodes parameter of "sample_infohashes" reply

### DISCUSSION THREAD

**Discussion by mattie20** [arvidn/libtorrent#7772] [Q&A]
I was wondering the amount of nodes libtorrent sends when it receives a sample_infohashes query. I can't seem to find it in the commit 219a8e7f6358e61a5d5d9646c5211dbc5f835495 the parameter defined in BEP 51 as "nodes". I thought it might be put in the dictionary empty or just left out all-together, but I have had a difficult time trying to track it down. I got as far as node.cpp and the corresponding function get_infohashes_sample() in dht_storage.cpp.

**Comment by arvidn** [comment]
the handler for that message is implemented here: https://github.com/arvidn/libtorrent/blob/master/src/kademlia/dht_storage.cpp#L457-L470

It looks like it's capped at 20 info-hashes

**Reply by mattie20** [reply]
"nodes" is a misnomer in that it is an array of peers that are sent along with the infohashes. I dug a little deeper following how the sample_infohash query is parsed. I originally thought that write_nodes_entries() put the packet in the sending queue, but it seems to add the additional "nodes" entry. The Number of nodes used seems to be the default value for the find_node function ( the default for count is 0 ). In the function when 0 it defaults to the bucket size. The bucket size is then some power of 2 between 0 and 256. So, the number of the peers included in the "nodes" parameter is between 0 and 256. Does this seem to be correct?
https://github.com/arvidn/libtorrent/blob/e5fe95f92273bd775961e2f1f3b4ec45f8012d9e/src/kademlia/node.cpp#L1159-L1182
https://github.com/arvidn/libtorrent/blob/60eead7e5a900cfb00a6e781e23525e612b75837/src/kademlia/node.cpp#L1186-L1193
https://github.com/arvidn/libtorrent/blob/e5fe95f92273bd775961e2f1f3b4ec45f8012d9e/src/kademlia/routing_table.cpp#L1103-L1107
